Implement support for LUKS-backed localstorage volumes

simu opened the pull request:
An encryption key can be specified per node in environment variable `ANSIBLE_OPENSHIFT_LOCALSTORAGE_<node_name>`, where `<node_name>` is constructed as `inventory_hostname.split('.')[0]`.

New settings fields `encrypted` (defaults to False, if unset) and `luks_key_url` (defaults to 'N/A' if unset) are introduced.
* `encrypted` controls whether the volume should use LUKS on LVM.
* `luks_key_url` can be used to point the user to a password safe entry where the key can be found. Omitting `luks_key_url` has no impact on the actual work done.

simu committed:
Implement support for LUKS-backed localstorage volumes

An encryption key can be specified per node in environment variable
`ANSIBLE_OPENSHIFT_LOCALSTORAGE_<node_name>`, where `<node_name>` is
constructed as `inventory_hostname.split('.')[0]`.

New settings fields `encrypted` (defaults to False, if unset) and
`luks_key_url` (defaults to 'N/A' if unset) are introduced.
`encrypted` controls whether the volume should use LUKS on LVM.
`luks_key_url` can be used to point the user to a password safe entry
where the key can be found. Omitting `luks_key_url` has no impact on the
actual work done.

simu committed:
Bump ansible-module-openshift dependency to v1.5.0

simu committed:
Use storage class "local-luks" as default for LUKS-backed volumes

Use storage class name "local-luks" for LUKS-backed volumes unless the
storage class name has been overridden for the volume.

mhutter commented:
I assume it's required for those volumes to be unlocked before the Kubelet starts? https://git.vshn.net/appuio/mungg-maint/-/merge_requests/78

simu commented:
> I assume it's required for those volumes to be unlocked before the Kubelet starts? https://git.vshn.net/appuio/mungg-maint/-/merge_requests/78

Yes, the idea is that `/usr/local/sbin/start-encrypted-devices` will be present on nodes hosting encrypted localstorage volumes (ensured via Puppet). That should cause the maintenance playbook to pause after a reboot until the volumes are unlocked (cf. https://git.vshn.net/vshn-ansible/capra-server-reboot/-/blob/master/tasks/post-reboot.yml#L79-103).

This still needs to be verified though.

mhutter commented:
But this in turn would require for `atomic-openshift-node` to be disabled and only started manually after `start-encrypted-devices` has run? :-/

simu commented:
> But this in turn would require for `atomic-openshift-node` to be disabled and only started manually after `start-encrypted-devices` has run? :-/

I don't think that's necessary, since no pods will be scheduled until the node is uncordoned which happens after the maintenance playbook pauses for the disk unlocks, and the Kubelet itself doesn't care about the `hostPath` PVs unless a pod is requesting them to be mounted as far as I can tell from the docs.

simu committed:
Add mount option nofail to encrypted localstorage volumes

simu committed:
Remove incorrect tempfile removal in cryptvol.yml

The LUKS key tempfile is now managed in tasks/main.yml and needs to
stick around until all volumes are created.

simu committed:
Deploy start-encrypted-devices script on nodes with encrypted volumes

simu committed:
Use keyfile in LUKS-encrypted file to unlock volumes

The keyfile and the volumes are encrypted using the same key, so in case
the LUKS container holding the keyfile gets lost, we can recreate the
keyfile using the key stored in the password manager.

simu commented:
The latest version of the feature deploys a script to unlock the encrypted volumes as `/usr/local/sbin/start-encrypted-volumes`.

Additionally we now use a LUKS container holding a keyfile to unlock the actual localstorage volumes. This change simplifies the procedure for unlocking the volume after the node has been rebooted, as with this change the encryption password needs to be entered only once (for the LUKS container holding the keyfile) instead of once per encrypted localstorage volume.

Repeated testing of the maintenance procedure has not shown any problems with localstorage volumes being unavailable when the Kubelet starts, as long as the volumes are unlocked and mounted before workloads using the volumes are scheduled on the node again. This condition is ensured as the mainenance pauses for unlocking the volumes before uncordoning the node, if the file `/usr/local/sbin/start-encrypted-volumes` is present on the node.

megian commented on README.md:
```diff
@@ -17,6 +17,11 @@ containers as `hostPath` persistent storage objects.
   * `pv_storage_class`: Storage class name
   * `pv_reclaim_policy`: Reclaim policy
+  * `encrypted`: Whether to setup the localstorage volume on a LUKS-on-LVM volume.
+     Encrypted volumes are skipped if the environment variable `ANSIBLE_OPENSHIFT_LOCALSTORAGE_<node>` is not set.
```
`Encrypted volumes are skipped, if`?

simu replied:
Comma is not required, cf. https://www.grammarly.com/blog/conditional-sentences/

> If the main clause precedes the if-clause, no punctuation is necessary.

megian replied:
Good to know.

megian commented on meta/main.yml:
```diff
@@ -7,4 +7,4 @@ galaxy_info:
 dependencies:
   - src: git+https://github.com/appuio/ansible-module-openshift.git
-    version: v1.3.4
+    version: v1.5.0
```
Even surprised, that this didn't crash before. Usage of the k8s module would be better, but I think this still works good enough.

simu replied:
ansible-galaxy complains about roles having mismatched versions of dependencies and picks one, I didn't check which one.

megian reviewed (approved):
Because the https://docs.ansible.com/ansible/2.9/modules/luks_device_module.html requires at least Ansible 2.8. We should rise min_ansible_version to 2.9 as well.

simu committed:
Bump `min_ansible_version` to 2.8

The luks_device module is only available with Ansible 2.8+.

simu commented:
Bumped `min_ansible_version` to 2.8, since that's the minimum required for `luks_device`.
